src/software/sensor_fusion/filter/ball_filter.h

```diff
@@ -52,25 +45,70 @@ class BallFilter
      */
-    explicit BallFilter(unsigned int min_buffer_size = DEFAULT_MIN_BUFFER_SIZE,
-                        unsigned int max_buffer_size = DEFAULT_MAX_BUFFER_SIZE);
+    explicit BallFilter();
 
     /**
-     * Filters the new ball detection data, and returns the updated state of the ball
-     * given the new data *
-     * @param current_ball_state The current state of the Ball
+     * Update the filter with the new ball detection data, and returns the new
+     * estimated state of the ball given the new data
+     *
      * @param new_ball_detections A list of new Ball detections
+     * @param filter_area The area within which the ball filter will work. Any detections
+     * outside of this area will be ignored.
      *
-     * @return The updated state of the ball given the new data
+     * @return The new estimated state of the ball given the new data. If a filtered
+     * result cannot be calculated, returns std::nullopt
+     */
+    std::optional<TimestampedBallState> estimateBallState(
+        const std::vector<BallDetection>& new_ball_detections,
+        const Rectangle& filter_area);
+
+   private:
+    /**
+     * A simple struct we use to pass around velocity estimate data
+     */
+    struct BallVelocityEstimate
+    {
+        Vector average_velocity;
+        double average_velocity_magnitude;
+        // The average of the max velocity magnitude and min velocity magnitude
+        double min_max_magnitude_average;
+    };
+
+    /**
+     * A simple struct to pass around linear regression data
      */
-    std::optional<TimestampedBallState> getFilteredData(
-        const std::vector<BallDetection> &new_ball_detections, const Field &field);
+    struct LinearRegressionResults
+    {
+        Line regression_line;
+        double regression_error;
+    };
+
+    /**
+     * Adds ball detections to the buffer stored by this filter. This function will ignore
+     * data if:
+     * - the data is outside of the filter_area, or
+     * - the data is too far away from the current known ball position
+     *   (since it is likely to be random noise).
+     *
+     * @param new_ball_detections The ball detections to try add to the buffer
+     * @param filter_area The area within which the ball filter will work. Any detections
+     * outside of this area will be ignored.
+     */
+    void addNewDetectionsToBuffer(std::vector<BallDetection> new_ball_detections,
+                                  const Rectangle& filter_area);
+
+    /**
+     * Uses linear regression to filter the given list of ball detections to find the
+     * current "real" state of the ball.
+     *
+     * @param ball_detections The detections to filter
+     *
+     * @return The filtered current state of the ball. if a filtered result cannot be
+     * calculated, returns std::nullopt
+     */
+    static std::optional<TimestampedBallState> estimateBallStateFromBuffer(
+        boost::circular_buffer<BallDetection> ball_detections);
```

Comment:
```suggestion
        const boost::circular_buffer<BallDetection> &ball_detections);
```

Reply:
Can't do because we resize the buffer (and sort it)